## Importing Data

In [1]:
from sklearn.preprocessing import OneHotEncoder, LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

from functions import *

In [2]:
# Dropping Features
bank_df.drop(columns=["Loan_ID", "Customer_ID"], inplace=True)

# Missing values
bank_df.isna().sum()

Loan_Status                     0
Current_Loan_Amount             0
Term                            0
Credit_Score                    0
Annual_Income                   0
Years_in_current_job            0
Home_Ownership                  0
Purpose                         0
Monthly_Debt                    0
Years_of_Credit_History         0
Months_since_last_delinquent    0
Number_of_Open_Accounts         0
Number_of_Credit_Problems       0
Current_Credit_Balance          0
Maximum_Open_Credit             0
Bankruptcies                    0
Tax_Liens                       0
dtype: int64

In [3]:
bank_df

Loan_Status  Current_Loan_Amount        Term  Credit_Score  \
2      Fully Paid           99999999.0  Short Term         741.0   
6      Fully Paid             217646.0  Short Term         730.0   
8      Fully Paid             548746.0  Short Term         678.0   
10     Fully Paid           99999999.0  Short Term         728.0   
12     Fully Paid           99999999.0  Short Term         740.0   
...           ...                  ...         ...           ...   
99990  Fully Paid           99999999.0  Short Term         742.0   
99993  Fully Paid              44484.0  Short Term         717.0   
99994  Fully Paid             210584.0  Short Term         719.0   
99996  Fully Paid           99999999.0  Short Term         732.0   
99997  Fully Paid             103136.0  Short Term         742.0   

       Annual_Income Years_in_current_job Home_Ownership             Purpose  \
2          2231892.0              8 years       Own Home  Debt Consolidation   
6          1184194.0             < 1 year  Home Mortgage  Debt Consolidation   
8          2559110.0              2 years           Rent  Debt Consolidation   
10          714628.0              3 years           Rent  Debt Consolidation   
12          776188.0             < 1 year       Own Home  Debt Consolidation   
...              ...                  ...            ...                 ...   
99990      1190046.0             < 1 year           Rent               other   
99993      1152426.0            10+ years  Home Mortgage      small_business   
99994       783389.0               1 year  Home Mortgage               Other   
99996      1289416.0               1 year           Rent  Debt Consolidation   
99997      1150545.0              6 years           Rent  Debt Consolidation   

       Monthly_Debt  Years_of_Credit_History  Months_since_last_delinquent  \
2          29200.53                     14.9                          29.0   
6          10855.08                     19.6                          10.0   
8          18660.28                     22.6                          33.0   
10         11851.06                     16.0                          76.0   
12         11578.22                      8.5                          25.0   
...             ...                      ...                           ...   
99990      11969.81                     20.1                          16.0   
99993       6280.64                     21.0                          12.0   
99994       3727.61                     17.4                          18.0   
99996      13109.05                      9.4                          21.0   
99997       7315.57                     18.8                          18.0   

       Number_of_Open_Accounts  Number_of_Credit_Problems  \
2                         18.0                        1.0   
6                         13.0                        1.0   
8                          4.0                        0.0   
10                        16.0                        0.0   
12                         6.0                        0.0   
...                        ...                        ...   
99990                      9.0                        0.0   
99993                      6.0                        0.0   
99994                      6.0                        0.0   
99996                     22.0                        0.0   
99997                     12.0                        1.0   

       Current_Credit_Balance  Maximum_Open_Credit  Bankruptcies  Tax_Liens  
2                    297996.0             750090.0           0.0        0.0  
6                    122170.0             272052.0           1.0        0.0  
8                    437171.0             555038.0           0.0        0.0  
10                   203965.0             289784.0           0.0        0.0  
12                   134083.0             220220.0           0.0        0.0  
...                       ...                  ...           ...        ...  
99990       

# Looking at Categorical Columns

## Purpose Column
There are many values in the Purpose column and will need to get rid of some values
- Values to get rid of:
    - other and Other
    - moving
    - wedding
    - vacation
    - Educational Expenses
    - renewable_energy
    
This means that values under 100 occurrences will be dropped

In [11]:
# Checking the Purpose column
bank_df.Purpose.value_counts()

Debt Consolidation      28202
Home Improvements        2409
other                    2313
Other                    1150
Business Loan             596
Medical Bills             444
Buy a Car                 428
Buy House                 263
Take a Trip               219
major_purchase            116
small_business            111
moving                     55
vacation                   41
wedding                    39
Educational Expenses       33
renewable_energy            4
Name: Purpose, dtype: int64

In [12]:
bank_df.Purpose.value_counts(normalize=True)

Debt Consolidation      0.774291
Home Improvements       0.066140
other                   0.063504
Other                   0.031573
Business Loan           0.016363
Medical Bills           0.012190
Buy a Car               0.011751
Buy House               0.007221
Take a Trip             0.006013
major_purchase          0.003185
small_business          0.003048
moving                  0.001510
vacation                0.001126
wedding                 0.001071
Educational Expenses    0.000906
renewable_energy        0.000110
Name: Purpose, dtype: float64

## Home Ownership Column

In [5]:
bank_df.Home_Ownership.value_counts()

Home Mortgage    18831
Rent             14300
Own Home          3205
HaveMortgage        87
Name: Home_Ownership, dtype: int64

## Years in Current Job Columns

In [6]:
bank_df.Years_in_current_job.value_counts()

10+ years    13149
2 years       3225
3 years       2997
< 1 year      2699
5 years       2487
4 years       2286
1 year        2247
6 years       2109
7 years       2082
8 years       1675
9 years       1467
Name: Years_in_current_job, dtype: int64

## Term Column

In [9]:
bank_df.Term.value_counts(normalize=True)

Short Term    0.709854
Long Term     0.290146
Name: Term, dtype: float64

## Dropping Irregular values

In [4]:
irregular_values = bank_df.loc[bank_df.Current_Loan_Amount == 99999999.0].index

bank_df.drop(index=irregular_values, inplace=True)

## Encoding
Performing One-hot-encoding to keep the interpretability

In [5]:
ohe = OneHotEncoder(drop="first")

In [6]:
# Getting data types that are objects
cat_var = bank_df.select_dtypes(include="object")
num_var = bank_df.select_dtypes(exclude="object")

In [7]:
# One hot encoding categorical variables
array_to_df = ohe.fit_transform(bank_df[cat_var.columns.tolist()]).toarray()  # Array values of the transformed columns

encoded = pd.DataFrame(array_to_df, columns=ohe.get_feature_names(cat_var.columns))  # Creating a pandas dataframe

bank_df = num_var.join(encoded, how="left")  # Combining the categorical variables and the numeric variables

In [8]:
bank_df.head()

Current_Loan_Amount  Credit_Score  Annual_Income  Monthly_Debt  \
6              217646.0         730.0      1184194.0      10855.08   
8              548746.0         678.0      2559110.0      18660.28   
14             234124.0         727.0       693234.0      14211.24   
17             666204.0         723.0      1821967.0      17612.24   
20             317108.0         687.0      1133274.0       9632.81   

    Years_of_Credit_History  Months_since_last_delinquent  \
6                      19.6                          10.0   
8                      22.6                          33.0   
14                     24.7                          46.0   
17                     22.0                          34.0   
20                     17.4                          53.0   

    Number_of_Open_Accounts  Number_of_Credit_Problems  \
6                      13.0                        1.0   
8                       4.0                        0.0   
14                     10.0                        1.0   
17                     15.0                        0.0   
20                      4.0                        0.0   

    Current_Credit_Balance  Maximum_Open_Credit  ...  Purpose_Medical Bills  \
6                 122170.0             272052.0  ...                    0.0   
8                 437171.0             555038.0  ...                    0.0   
14                 28291.0             107052.0  ...                    0.0   
17                813694.0            2004618.0  ...                    0.0   
20                 60287.0             126940.0  ...                    0.0   

    Purpose_Other  Purpose_Take a Trip  Purpose_major_purchase  \
6             0.0                  0.0                     0.0   
8             0.0                  0.0                     0.0   
14            0.0                  0.0                     0.0   
17            0.0                  0.0                     0.0   
20            0.0                  0.0                     0.0   

    Purpose_moving  Purpose_other  Purpose_renewable_energy  \
6              0.0            0.0                       0.0   
8              0.0            0.0                       0.0   
14             0.0            0.0                       0.0   
17             0.0            0.0                       0.0   
20             0.0            0.0                       0.0   

    Purpose_small_business  Purpose_vacation  Purpose_wedding  
6                      0.0               0.0              0.0  
8                      0.0               0.0              0.0  
14                     0.0               0.0              0.0  
17                     0.0               0.0              0.0  
20                     0.0               0.0              0.0  

[5 rows x 42 columns]

In [9]:
bank_df.shape

(31505, 42)

In [10]:
bank_df.columns

Index(['Current_Loan_Amount', 'Credit_Score', 'Annual_Income', 'Monthly_Debt',
       'Years_of_Credit_History', 'Months_since_last_delinquent',
       'Number_of_Open_Accounts', 'Number_of_Credit_Problems',
       'Current_Credit_Balance', 'Maximum_Open_Credit', 'Bankruptcies',
       'Tax_Liens', 'Loan_Status_Fully Paid', 'Term_Short Term',
       'Years_in_current_job_10+ years', 'Years_in_current_job_2 years',
       'Years_in_current_job_3 years', 'Years_in_current_job_4 years',
       'Years_in_current_job_5 years', 'Years_in_current_job_6 years',
       'Years_in_current_job_7 years', 'Years_in_current_job_8 years',
       'Years_in_current_job_9 years', 'Years_in_current_job_< 1 year',
       'Home_Ownership_Home Mortgage', 'Home_Ownership_Own Home',
       'Home_Ownership_Rent', 'Purpose_Buy House', 'Purpose_Buy a Car',
       'Purpose_Debt Consolidation', 'Purpose_Educational Expenses',
       'Purpose_Home Improvements', 'Purpose_Medical Bills', 'Purpose_Other',
       'Purp

## Training and Testing Sets

In [11]:
y = bank_df["Loan_Status_Fully Paid"]
X = bank_df.drop(columns=["Loan_Status_Fully Paid"])

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=3, test_size=0.2)

In [12]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((25204, 41), (6301, 41), (25204,), (6301,))

In [13]:
X_train.head()

Current_Loan_Amount  Credit_Score  Annual_Income  Monthly_Debt  \
55605             669768.0         734.0      3856240.0      14750.08   
40511             198242.0         734.0       951140.0      22272.37   
46486             291918.0         747.0       799121.0      10921.20   
26147             348018.0         726.0      2254236.0      59173.79   
83792             228448.0        7350.0      1721438.0      40166.95   

       Years_of_Credit_History  Months_since_last_delinquent  \
55605                     32.7                          48.0   
40511                     23.5                          17.0   
46486                     13.7                          41.0   
26147                     25.0                          11.0   
83792                     21.5                          34.0   

       Number_of_Open_Accounts  Number_of_Credit_Problems  \
55605                      8.0                        0.0   
40511                     11.0                        0.0   
46486                     12.0                        0.0   
26147                     12.0                        0.0   
83792                     15.0                        0.0   

       Current_Credit_Balance  Maximum_Open_Credit  ...  \
55605                331968.0             591360.0  ...   
40511                367403.0             743754.0  ...   
46486                167523.0             629816.0  ...   
26147                584953.0             781220.0  ...   
83792                278749.0             461120.0  ...   

       Purpose_Medical Bills  Purpose_Other  Purpose_Take a Trip  \
55605                    NaN            NaN                  NaN   
40511                    NaN            NaN                  NaN   
46486                    NaN            NaN                  NaN   
26147                    0.0            0.0                  0.0   
83792                    NaN            NaN                  NaN   

       Purpose_major_purchase  Purpose_moving  Purpose_other  \
55605                     NaN             NaN            NaN   
40511                     NaN             NaN            NaN   
46486                     NaN             NaN            NaN   
26147                     0.0             0.0            0.0   
83792                     NaN             NaN            NaN   

       Purpose_renewable_energy  Purpose_small_business  Purpose_vacation  \
55605                       NaN                     NaN               NaN   
40511                       NaN                     NaN               NaN   
46486                       NaN                     NaN               NaN   
26147                       0.0                     0.0               0.0   
83792                       NaN                     NaN               NaN   

       Purpose_wedding  
55605              NaN  
40511              NaN  
46486              NaN  
26147              0.0  
83792              NaN  

[5 rows x 41 columns]

## Missing Values
There are a lot of missing values in the new encoded values

In [15]:
# Missing values
X_train.isna().sum().sum()

504948

In [16]:
X_train.dropna(inplace=True)

<ipython-input-16-c608130e58c9>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_train.dropna(inplace=True)


In [18]:
X_train.shape

(7792, 41)

## Observation
After cleaning the entire dataset we are left with 7,792 rows of data from the original 100,000+